### ISSUE TITLE
Unexpected exception for classes having a ReadOnlySpan<> constructor

### ISSUE THREAD

**Issue by oblaise** [dotnet/command-line-api#1137]
Dear,

Since I upgraded to the latest version of the nuget package (2.0.0-beta1.20574.7), coming from previous version (2.0.0-beta1.20371.2).
I'm facing an exception when I'm not specifying some options on the argument line. The previous package didn't cause that exception to be thrown. 
Apparently this exception occurs as I've options that use `IPAddress` type for which I use a `parseArgument` function but `IPAddress` has a constructor with a `ReadOnlySpan<byte>` argument that apparently causes the exception when the option isn't specified in the command line.

Here is a small .NET 5 program that reproduces the issue:
```
using System;
using System.Net;
using System.CommandLine;
using System.CommandLine.Binding;
using System.CommandLine.Parsing;
using System.CommandLine.Invocation;

#nullable enable

RootCommand root = new RootCommand()
{
    new Option<IPAddress?>("--host", parseArgument: ConvertIPAddress, description: "IP address of the server"),
    new Option<ushort>("--port"),
};
root.Handler = CommandHandler.Create<ConnectionOptions>(options=> {Console.WriteLine($"Args: {options.Host}:{options.Port}");} );

root.Invoke("--host 2.3.4.5 --port 5678 ");
root.Invoke("--host 2.3.4.5");
// throw exception System.InvalidOperationException: The type System.ReadOnlySpan`1[System.Byte] cannot be bound
root.Invoke("--port 5678");

static IPAddress? ConvertIPAddress(ArgumentResult result)
{
    if (result.Tokens.Count == 1)
    {
        if (IPAddress.TryParse(result.Tokens[0].Value, out IPAddress? outAddress))
            return outAddress;
        result.ErrorMessage=$"Invalid IPAddress : {result.Tokens[0].Value}";
        return null;
    }
    return IPAddress.Any;
}

public class ConnectionOptions
{
    public IPAddress? Host { get; set; } = null;
    public ushort Port { get; set; } = 1234;
}

```

**Comment by oblaise** [comment]
After further analysis and tests there is no link to the parseArgument functionnality. I chaged the title of the Issue.

Using a type that has two single arguments constructors, one using `string`, one using `ReadOnlySpan<>` triggers the exception.

Following code cause an exception:
```
using System;
using System.CommandLine;
using System.CommandLine.Parsing;
using System.CommandLine.Invocation;

RootCommand root = new RootCommand()
{
    new Option<MyGuid>("--guid"),
    new Option<int>("--value"),
};
root.Handler = CommandHandler.Create<MyGuid,int>((guid,value)=> {Console.WriteLine($"Guid: {guid}, Value: {value}");} );
root.Invoke("--guid 5c4f219f-59a6-4820-ba0a-61ec270a2000 --value 1234 ");
root.Invoke("--guid 5c4f219f-59a6-4820-ba0a-61ec270a2000");
// The following call trigger exception
root.Invoke("--value 1234");

public class MyGuid
{
    private Guid value;
    public MyGuid(string guid) {  value=new Guid(guid); }
    public MyGuid(ReadOnlySpan<byte> guid) { value = new Guid(guid);}
    public override string ToString() => value.ToString();
}
```

Using `Guid?` as type instead of `MyGuid` in the above code doesn't triggers the exception.
Using `Guid` triggers another exception. Is it normal that the use of a struct doesn't use `default` value when the option isn't specified but instead trigger an exception ?

**Comment by iesoftwaredeveloper** [comment]
I receive this exception as well.  I am able to reproduce it when I am trying to use use an optional decimal and ComplexTypes.  This doesn't seem to occur if I do not use ComplexTypes.

I am using System.CommandLine 2.0.0-beta1.21215.1
Target Framework is net5.0

Providing the option on the command line is successful.

```
$ dotnet run -- --opt-decimal 123.456
ParseResult: [ system-commandline-decimal-exception [ --opt-decimal <123.456> ] ]
--opt-decimal 123.456
```

Not including the optional parameter on the command line results in the exception.

```
$ dotnet run
Unhandled exception: System.InvalidOperationException: The type System.ReadOnlySpan`1[System.Int32] cannot be bound
   at System.CommandLine.Binding.ModelBinder.CreateInstanceInternal(BindingContext bindingContext)
   at System.CommandLine.Binding.ModelBinder.GetBoundValue(IValueSource valueSource, BindingContext bindingContext, IValueDescriptor valueDescriptor, Boolean includeMissingValues, Type parentType)
   at System.CommandLine.Binding.ModelBinder.GetBoundValues(IDictionary`2 bindingSources, BindingContext bindingContext, IReadOnlyList`1 valueDescriptors, Boolean enforceExplicitBinding, Type parentType, Boolean includeMissingValues)
   at System.CommandLine.Binding.ModelBinder.GetBestConstructorAndArgs(BindingContext bindingContext)
   at System.CommandLine.Binding.ModelBinder.CreateInstanceInternal(BindingContext bindingContext)
   at System.CommandLine.Binding.ModelBinder.GetBoundValue(IValueSource valueSource, BindingContext bindingContext, IValueDescriptor valueDescriptor, Boolean includeMissingValues, Type parentType)
   at System.CommandLine.Binding.ModelBinder.GetBoundValues(IDictionary`2 bindingSources, BindingContext bindingContext, IReadOnlyList`1 valueDescriptors, Boolean enforceExplicitBinding, Type parentType, Boolean includeMissingValues)
   at System.CommandLine.Binding.ModelBinder.UpdateInstanceInternalNotifyIfNonDefaultsUsed[T](T instance, BindingContext bindingContext)
   at System.CommandLine.Binding.ModelBinder.InstanceFromSpecificConstructor(BindingContext bindingContext, ConstructorDescriptor constructor, IReadOnlyList`1 boundValues, Boolean& nonDefaultsUsed)
   at System.CommandLine.Binding.ModelBinder.CreateInstanceInternal(BindingContext bindingContext)
   at System.CommandLine.Binding.ModelBinder.GetBoundValue(IValueSource valueSource, BindingContext bindingContext, IValueDescriptor valueDescriptor, Boolean includeMissingValues, Type parentType)
   at System.CommandLine.Binding.ModelBinder.GetBoundValues(IDictionary`2 bindingSources, BindingContext bindingContext, IReadOnlyList`1 valueDescriptors, Boolean enforceExplicitBinding, Type parentType, Boolean includeMissingValues)
   at System.CommandLine.Invocation.ModelBindingCommandHandler.InvokeAsync(InvocationContext context)
   at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<<BuildInvocationChain>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass23_0.<<UseParseErrorReporting>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass16_0.<<UseHelp>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass27_0.<<UseVersionOption>b__1>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass25_0.<<UseTypoCorrections>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseSuggestDirective>b__24_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass22_0.<<UseParseDirective>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass11_0.<<UseDebugDirective>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<RegisterWithDotnetSuggest>b__10_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass14_0.<<UseExceptionHandler>b__0>d.MoveNext()
```

Here is a quick sample program to reproduce.
```
using System.CommandLine;
using System.CommandLine.Invocation;
using System.CommandLine.IO;
using System.CommandLine.Parsing;

namespace system_commandline_decimal_exception
{
    /// <summary>
    /// Successful example
    /// 
    /// $ dotnet run -- --opt-decimal 123.456
    /// ParseResult: [ system-commandline-decimal-exception [ --opt-decimal <123.456> ] ]
    /// --opt-decimal 123.456
    /// 
    /// Exception
    /// 
    /// $ dotnet run -- 
    /// </summary>
    class Program
    {
        static void Main(string[] args)
        {
            var rootCommand = new RootCommand()
            {
                Description = "Console app to demonstrate System.CommandLine complex type exception"
            };
            rootCommand.Add(new Option<decimal>(new[] { "--opt-decimal" }, description: "A decimal option"));
            rootCommand.Handler = CommandHandler.Create((ParseResult parseResult, IConsole console, ComplexType options) =>
               {
                   console.Out.WriteLine($"{parseResult}");
                   console.Out.WriteLine($"--opt-decimal {options.OptDecimal}");
               });

            rootCommand.InvokeAsync(args);
        }
    }

    public class ComplexType
    {
        public decimal OptDecimal { get; set; }
    }

}
```

**Comment by oblaise** [comment]
Dear @jonsequitur,

This solved the linked issue for decimal parsing #1284 but doesn't seems to solve the original #1137 issue as far as my tests described above are concerned.

**Comment by elgonzo** [comment]
I can confirm that there still is an exception regarding `ReadOnlySpan<byte>` and i would like to see the issue being reopened.

While it makes sense that ReadOnlySpan is not supported by System.CommanLine, the library stumbles when trying to find a matching constructor in the absence of an option value when confronted with a class like MyGuid from oblaise's example above.

The problem seems to lie in the method ModelBinder.GetBestConstructorAndArgs: https://github.com/dotnet/command-line-api/blob/3264927b51a5efda4f612c3c08ea1fc089f4fc35/src/System.CommandLine/Binding/ModelBinder.cs#L165-L197

It iterates over all available constructors until it finds the best matching constructor, otherwise it will chose the first "bestNonMatching" constructor found.

And here the problem manifests. In the absence of the option value, the constructor with the string parameter in the MyGuid class from the example by @oblaise above is only found as a "bestNonMatching" constructor. In that situation, the GetBestConstructorAndArgs method keeps inspecting any other constructors in an attempt to find a perfectly matching constructor. And it eventually runs into the constructor with the ReadOnlySpan<>, at which point it throws an exception due to ReadOnlySpan<> not being supported. Kaboom!

I am not sure what the solution to this issue would be. Instinctively i would tend towards catching any exception in GetBestConstructorAndArgs, and only rethrowing the first or last caught exception if there has been no matching constructor found (e.g., GetBestConstructorAndArgs would return null).
